### ISSUE TITLE
Remote sessions targeting Windows machines unexpectedly use ANSI character encoding for external programs

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#15128]
Note: It is debatable what console encoding should apply when remoting into a Window machine - the caller's or the target machine's - but (effectively) using _ANSI_ encoding  - rather than the OEM code page, as is still the default on Windows - seems like the wrong choice.

Curiously, Windows PowerShell behaves this way too.

## Steps to reproduce

Run on (PS Core) on Windows:

* from an _elevated_ session, to enable "loopback" remoting
* make sure that the calling version is enabled for remoting 

```powershell
& {
  # Target the local machine.
  $htParams = @{ ('HostName', 'ComputerName')[$env:OS -eq 'Windows_NT'] = '.' }
  # On Windows, determine what edition / version to target remotely: use the caller's.
  if ($env:OS -eq 'Windows_NT') {
    $htParams['ConfigurationName'] = (('PowerShell.' + $PSVersionTable.PSVersion), 'microsoft.powershell')[$PSVersionTable.PSEdition -ne 'Core']
  }  
  Invoke-Command @htParams { if ($env:OS -eq 'Windows_NT') { chcp } else { $env:LANG }; [Console]::InputEncoding.WebName; [Console]::OutputEncoding.WebName, $OutputEncoding.WebName }
}
```

## Expected behavior

```powershell
Active code page: 437
ibm437   # [Console]::InputEncoding.WebName
ibm437   # [Console]::OutputEncoding.WebName
utf-8       # $OutputEncoding
```

## Actual behavior

```powershell
Active code page: 437
windows-1252   # !! *ANSI* code page
Codepage - 0   # !! *0* - but in effect seems to behave like ANSI
utf-8
```

## Environment data

```none
PowerShell Core 7.2.0-preview.4
```

**Comment by jborean93** [comment]
While the behaviour definitely isn't helpful the reason why this is the case is because your PSSession does not have a console/conhost.exe allocated to the process. When .NET goes to get the codepage [GetConsoleCP](https://docs.microsoft.com/en-us/windows/console/getconsolecp) (Input) and [GetConsoleOutputCP](https://docs.microsoft.com/en-us/windows/console/getconsoleoutputcp) (Output) both return 0 meaning an error occurred.

This value isn't checked at all and is just passed straight into `[Text.Encoding]::GetEncoding` where the docs state https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.getencoding?view=net-5.0#System_Text_Encoding_GetEncoding_System_Int32_

> 0 (zero), to use the default encoding.

The reason why running `chcp.com` returns the correct codepage is because the act of starting a spawning a new console process without an existing conhost will result in one being spawned automatically and it now has the correct value.

Try it out

```powershell
Invoke-Command -ComputerName localhost -ScriptBlock {

    $getConsoleCP = {
        Add-Type -Namespace Kernel32 -Name NativeMethods -MemberDefinition @'
[DllImport("Kernel32.dll")]
public static extern uint GetConsoleCP();

[DllImport("Kernel32.dll")]
public static extern uint GetConsoleOutputCP();
'@

        [PSCustomObject]@{
            ConsoleCP = [Kernel32.NativeMethods]::GetConsoleCP()
            InputEncoding = [Console]::InputEncoding.WebName

            ConsoleOutputCP = [Kernel32.NativeMethods]::GetConsoleOutputCP()
            OutputEncoding = [Console]::OutputEncoding.WebName
        }
    }
    
    Write-Host "In PSSession"
    &$getConsoleCP | Format-Table

    Write-Host "In sub process"
    $encCommand = [Convert]::ToBase64String([Text.Encoding]::Unicode.GetBytes($getConsoleCP.ToString()))
    powershell.exe -EncodedCommand $encCommand
}
```

Outputs

```
In PSSession

ConsoleCP InputEncoding ConsoleOutputCP OutputEncoding
--------- ------------- --------------- --------------
        0 Windows-1252                0 Windows-1252


In sub process

ConsoleCP InputEncoding ConsoleOutputCP OutputEncoding
--------- ------------- --------------- --------------
      437 IBM437                    437 IBM437
```

**Comment by jborean93** [comment]
Now the question is what should be the correct behaviour, some of the options I can think off right now

* .NET fixes their logic to fallback to the registry setting if the Win32 APIs return 0
* PowerShell first checks if a console is present
  * If present then it acts like it does today
  * If not it could fallback to the registry
* PowerShell could automatically allocate a new conhost whenever a native process is started and set the cp to 65001 (or fallback to the registry default)
  * If PowerShell spawns a console then any child processes it creates should use that same console/cp settings

For the latter it will have the same restrictions that exist today for a normal process session but it at least solves this problem.

**Comment by mklement0** [comment]
Thanks, @jborean93, that's helpful background information.

I explicitly included the `chcp` call in order to force allocation of a console first, followed by querying `[Console]::InputEncoding` and `[Console]::OutputEncoding`.

The curious thing that the `chcp` calls assigns a value to `[Console]::InputEncoding`, it doesn't do so for `[Console]::OutputEncoding`

By contrast, in _Windows PowerShell_ _both_ values are set to the ANSI encoding, and even _without_ running a console application first - does it always, automatically allocate a console?

**Comment by mklement0** [comment]
As for what the correct behavior should be:

* .NET should definitely fix their fallback behavior to fall back to the _OEM_ code page, not the ANSI code page.

* Separately, we should decide what makes more sense for PowerShell remoting: inherit the _caller_'s code page, or defer to the targeted machine's code page? 
  * If the latter case, `65001` (UTF-8) would certainly be convenient, but only appropriate if we officially we decide to make PowerShell Core sessions on Windows UTF-8 _by default_, as proposed in https://github.com/PowerShell/PowerShell/issues/14945#issuecomment-800503184; otherwise, the target machine's OEM code page should be used.

**Comment by jborean93** [comment]
> I explicitly included the chcp call in order to force allocation of a console first, followed by querying [Console]::InputEncoding and [Console]::OutputEncoding.

The conhost is created and tied to the lifetime of the `chcp` process. It won't persist beyond that life of that process and tie itself to your PSSession so that codepage will still be 0. So regardless of the order the fact that you called `chcp` won't affect (or at least I don't see how) the .NET Console encoding values of your PSSession itself. The same thing applies to any other console subprocess you spawn. For example I've got a blank PSSession open I can see it's running as `wsmprovhost.exe` with no `conhost.exe` attached.

![image](https://user-images.githubusercontent.com/8462645/113087403-e0a11e00-9226-11eb-8796-fdbf11f849c9.png)

I spawn `powershell.exe` with a lifetime time to keep it running and I can see it now has a console associated with it

![image](https://user-images.githubusercontent.com/8462645/113087476-04fcfa80-9227-11eb-85d7-fbc7fb7b6984.png)

When that ends I'm now back with just `wsmprovhost.exe` with the `conhost.exe` no longer there

![image](https://user-images.githubusercontent.com/8462645/113087591-53aa9480-9227-11eb-8cb5-799742c4b23f.png)

> The curious thing that the chcp calls assigns a value to [Console]::InputEncoding, but not to [Console]::OutputEncoding

This is what doesn't make sense to me, both `InputEncoding` and `OutputEncoding` depend on the Win32 APIs to get the values. They both should be returning `0` and thus set the encoding to the default encoding on the host. Even when I test that assumption I cannot replicate what you see (even when targeting PowerShell v7).

![image](https://user-images.githubusercontent.com/8462645/113087718-94a2a900-9227-11eb-9495-e736a07d1ff2.png)

> By contrast, in Windows PowerShell both values are set to the ANSI encoding, and even without running a console application first - does it always, automatically allocate a console? Or does it simply take an extra step to translate 0 into the default == ANSI code page?

I'm not 100% sure what you mean with Windows PowerShell here, the encoding values come from .NET and it's the behaviour of `[Text.Encoding]::GetEncoding(0)` to return the default encoding (whatever that actually is) which in this case is the "ANSI" encoding.

**Comment by mklement0** [comment]
> The conhost is created and tied to the lifetime of the chcp process. It won't persist beyond that life of that process and tie itself to your PSSession so that codepage will still be 0. So regardless of the order the fact that you called chcp won't affect (or at least I don't see how) the .NET Console encoding values of your PSSession itself. 

Good to know, thanks. Yes, it is curious thing that (only) `[Console]::InputEncoding` is set after running `chcp`, but the fact that  reported code page _differs_ from what `chcp` reports confirms your larger point that `chcp`'s console is no longer active.

> I'm not 100% sure what you mean with Windows PowerShell here

Run the same command from Windows PowerShell, which targets itself (configuration `microsoft.powershell`) as the endpoint, and you'll see the difference.
It may come down to the difference between .NET Framework (Windows PowerShell) and .NET (Core) 5+ (PowerShell Core).

The above applies to the `Invoke-Command` invocation scenario; when entering an _interactive_ session with `Enter-PSSession`, I do consistently see both `[Console]::InputEncoding`  and `[Console]::OutputEncoding`  set to the ANSI encoding.

To simplify the `Invoke-Command` calls and show the difference (it doesn't make a difference what edition you call _from_):

```powershell
# Target the current PS Core preview (I get the same behavior with 7.0.3)
PS> icm -cn . -config PowerShell.7.2.0-preview.4 { [console]::InputEncoding.CodePage; [console]::OutputEncoding.CodePage }
0
0
```

```powershell
# Target WinPS
PS> icm -cn . -config microsoft.powershell { [console]::InputEncoding.CodePage; [console]::OutputEncoding.CodePage }
1252
1252
```

**Comment by jborean93** [comment]
> The above applies to the Invoke-Command invocation scenario; when entering an interactive session with Enter-PSSession, I do consistently see both [Console]::InputEncoding and [Console]::OutputEncoding set to the ANSI encoding.

Ah ok, I'm running with just `Enter-PSSession`, weird that there is a difference though, I wouldn't have thought it would matter.

**Comment by jborean93** [comment]
I'm currently looking into how `[Text.Encoding]::GetEncoding(0)` returns windows-1252. The definition of this function for .NET Core is at https://github.com/dotnet/runtime/blob/99d6215ef2336b8fa51fe19b2e8e782df62959df/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs#L215. Briefly browsing the logic is somewhat similar to .NET Framework but not exactly the same

It first checks if the encoding is in the registered provider (more on that later on) and then falls back to a hardcoded switch statement. In that statement `0` is set to be `[Text.Encoding]::Default` which in .NET Core is set to [UTF-8 without a BOM](https://github.com/dotnet/runtime/blob/99d6215ef2336b8fa51fe19b2e8e782df62959df/src/libraries/System.Private.CoreLib/src/System/Text/Encoding.cs#L80-L83). On .NET Framework [the `Default` encoding](https://referencesource.microsoft.com/#mscorlib/system/text/encoding.cs,da50c22465aa9274,references) it is dependent on [GetACP](https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-getacp).

So this explains why `[Text.Encoding]::GetEncoding(0)` returns `windows-1252` on Windows PowerShell but not PowerShell (I would expect that to be `utf-8`). In .NET Core there has to be a registered encoding provider that supports `0` overriding the default behaviour of `0` being `[Text.Encoding]::Default`.

We can use reflection to test this out on PowerShell:

```powershell
$encProvType = [AppDomain]::CurrentDomain.GetAssemblies().DefinedTypes | Where-Object {
    $_.FullName -eq 'System.Text.EncodingProvider'
} | Select -First 1

$getMeth = $encProvType.GetMethod('GetEncodingFromProvider', [Reflection.BindingFlags]'NonPublic, Static', $null, [Type[]]@([int]), $null)
$getMeth.Invoke($null, @(0))
```

_Outputs_

```
IsSingleByte      : True
EncodingName      : Western European (Windows)
WebName           : windows-1252
HeaderName        : windows-1252
BodyName          : iso-8859-1
Preamble          :
WindowsCodePage   :
IsBrowserDisplay  :
IsBrowserSave     :
IsMailNewsDisplay :
IsMailNewsSave    :
EncoderFallback   : System.Text.InternalEncoderBestFitFallback
DecoderFallback   : System.Text.InternalDecoderBestFitFallback
IsReadOnly        : True
CodePage          : 1252
```

So this explains why `GetEncoding(0)` returns `windows-1252`. When looking at the [GetEncodingFromProvider method](https://github.com/dotnet/runtime/blob/99d6215ef2336b8fa51fe19b2e8e782df62959df/src/libraries/System.Private.CoreLib/src/System/Text/EncodingProvider.cs#L83) the logic is pretty simple; it looks through the registered providers (`s_providers`) and returns the first match it finds for the codepage.

We can also use reflection to get these registered providers

```powershell
$encProvType = [AppDomain]::CurrentDomain.GetAssemblies().DefinedTypes | Where-Object {
    $_.FullName -eq 'System.Text.EncodingProvider'
} | Select -First 1

$sProvidersField = $encProvType.GetField('s_providers', [Reflection.BindingFlags]'NonPublic, Static')
$providers = [Text.EncodingProvider[]]$sProvidersField.GetValue($null)

$providers | % GetType
```

_Outputs_

```
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     False    CodePagesEncodingProvider                System.Text.EncodingProvider
```

The docs for the [CodePagesEncodingProvider](https://docs.microsoft.com/en-us/dotnet/api/system.text.codepagesencodingprovider?view=net-5.0) state it

> Provides access to an encoding provider for code pages that otherwise are available only in the desktop .NET Framework.

Looking at the PowerShell codebase it looks like it is being registered there https://github.com/PowerShell/PowerShell/blob/3ba61069b8e8d027dbbe11f69a85c871359291f9/src/System.Management.Automation/utils/ClrFacade.cs#L143-L149. I just don't have much of an idea of what the purpose of this particular code is for. It looks like some sort of shim to implement some .NET Framework behaviour that is missing in .NET Core. This could even be a red herring and not at all related to the problem but I thought it interesting enough to share.

**Comment by jborean93** [comment]
Alas I have no idea why `Invoke-Command` differs from `Enter-PSSession` here, they both don't have a console associated with the remote process so the raw Win32 values make sense but I don't know why PowerShell returns 0 (I'm seeing it on OutputEncoding on my host)
```powershell
$getConsoleCP = {
    param ([string]$Scenario, [bool]$RunChcp)
    
    if ($RunChcp) {
        chcp.com
    }

    Add-Type -Namespace Kernel32 -Name NativeMethods -MemberDefinition @'
[DllImport("Kernel32.dll")]
public static extern uint GetConsoleCP();

[DllImport("Kernel32.dll")]
public static extern uint GetConsoleOutputCP();
'@

    $inputEncoding = [Console]::InputEncoding
    $outputEncoding = [Console]::OutputEncoding

    [PSCustomObject]@{
        Scenario = $Scenario
        ConsoleCP = [Kernel32.NativeMethods]::GetConsoleCP()
        InputEncodingCP = $inputEncoding.CodePage
        InputEncoding = $inputEncoding.WebName
        
        
        ConsoleOutputCP = [Kernel32.NativeMethods]::GetConsoleOutputCP()
        OutputEncodingCP = $outputEncoding.CodePage
        OutputEncoding = $outputEncoding.WebName
    }
}

Invoke-Command localhost -ScriptBlock $getConsoleCP -ArgumentList 'WinPS', $false

Invoke-Command localhost -ScriptBlock $getConsoleCP -ArgumentList 'WinPS + proc', $true

Invoke-Command localhost -ConfigurationName PowerShell.7 -ScriptBlock $getConsoleCP -ArgumentList 'PS', $false

Invoke-Command localhost -ConfigurationName PowerShell.7 -ScriptBlock $getConsoleCP -ArgumentList 'PS + proc', $true
```

```
Scenario         : WinPS
ConsoleCP        : 0
InputEncodingCP  : 1252
InputEncoding    : Windows-1252
ConsoleOutputCP  : 0
OutputEncodingCP : 1252
OutputEncoding   : Windows-1252


Scenario         : WinPS + proc
ConsoleCP        : 0
InputEncodingCP  : 1252
InputEncoding    : Windows-1252
ConsoleOutputCP  : 0
OutputEncodingCP : 1252
OutputEncoding   : Windows-1252


Scenario         : PS
ConsoleCP        : 0
InputEncodingCP  : 1252
InputEncoding    : windows-1252
ConsoleOutputCP  : 0
OutputEncodingCP : 1252
OutputEncoding   : windows-1252


Scenario         : PS + proc
ConsoleCP        : 0
InputEncodingCP  : 1252
InputEncoding    : windows-1252
ConsoleOutputCP  : 0
OutputEncodingCP : 0
OutputEncoding   : Codepage - 0
```

WinPS is consistent and makes sense because it's really just using `[Text.Encoding]::Default` when the cp returned by the Win32 API is `0`. Pwsh is interesting as it uses the registered `CodePagesEncodingProvider` to get the "ANSI" codepage but when a console process had run that provider seems to be a bit screwed up and only for `[Console]::OutputEncoding`.

Maybe PSReadLine or some other mechanism is setting `[Console]::OutputEncoding = $someBackupValue` after the native process had exited and `$someBackupValue` was derived in some other way from `0`.

**Comment by iSazonov** [comment]
> As for what the correct behavior should be:
> 
> * .NET should definitely fix their fallback behavior to fall back to the _OEM_ code page, not the ANSI code page.

Is there an issue for reference?

> * Separately, we should decide what makes more sense for PowerShell remoting: inherit the _caller_'s code page, or defer to the targeted machine's code page?
>   
>   * If the latter case, `65001` (UTF-8) would certainly be convenient, but only appropriate if we officially we decide to make PowerShell Core sessions on Windows UTF-8 _by default_, as proposed in [#14945 (comment)](https://github.com/PowerShell/PowerShell/issues/14945#issuecomment-800503184); otherwise, the target machine's OEM code page should be used.

Of course, I would like to avoid a situation when we receive a message from a remote host in a different language and in wrong double re-encoding.

If we connect to WinPS we should follow old behavior. 
If we connect to new PowerShell Core it is best to be Utf8 (maybe excluding old Windows) as it will cause fewer side effects and reflects the modern trend, and it would nice to have a parameter to set explicitly encoding to address specific scenarios.

**Comment by iSazonov** [comment]
> This is all happening on the remote side,

I say about another - if our local session has English  locale and remote one - Russian, then we see follow:
```powershell
 Invoke-Command -ComputerName titlin -ScriptBlock { ping localhost }

ЋЎ¬Ґ­ Ї ЄҐв ¬Ё б testhost [::1] б 32 Ў ©в ¬Ё ¤ ­­ле:
ЋвўҐв ®в ::1: ўаҐ¬п<1¬б
ЋвўҐв ®в ::1: ўаҐ¬п<1¬б
ЋвўҐв ®в ::1: ўаҐ¬п<1¬б
ЋвўҐв ®в ::1: ўаҐ¬п<1¬б

‘в вЁбвЁЄ  Ping ¤«п ::1:
    Џ ЄҐв®ў: ®вЇа ў«Ґ­® = 4, Ї®«гзҐ­® = 4, Ї®вҐап­® = 0
    (0% Ї®вҐам)
ЏаЁЎ«Ё§ЁвҐ«м­®Ґ ўаҐ¬п ЇаЁҐ¬ -ЇҐаҐ¤ зЁ ў ¬б:
    ЊЁ­Ё¬ «м­®Ґ = 0¬бҐЄ, Њ ЄбЁ¬ «м­®Ґ = 0 ¬бҐЄ, ‘аҐ¤­ҐҐ = 0 ¬бҐЄ
```
Expected:
```powershell
Обмен пакетами с testhost [::1] с 32 байтами данных:
Ответ от ::1: время<1мс
Ответ от ::1: время<1мс
Ответ от ::1: время<1мс
Ответ от ::1: время<1мс


Статистика Ping для ::1:
    Пакетов: отправлено = 4, получено = 4, потеряно = 0
    (0% потерь)
Приблизительное время приема-передачи в мс:
    Минимальное = 0мсек, Максимальное = 0 мсек, Среднее = 0 мсек
ping.exe exited on titlin with error code 0.
```

**Comment by jborean93** [comment]
That’s most likely due to the same problem but had nothing to do with the PSRemoting serialisation. At that point it’s already a string and damage is already done. It could be that the culture settings that are sent through impact the default encoding used but it’s just more reason why PowerShell needs to be more defensive when there’s no console present.

**Comment by mklement0** [comment]
Great sleuthing, @jborean93.

@iSazonov, re:

> Is there an issue for reference?

I couldn't find anything (except an ancient issue, which shows that at some point the default _with_ an allocated console was UTF-8(!) - https://github.com/dotnet/runtime/issues/17849).

When accessing  `Console.In/OutputEncoding` _without a console present_:

* It is only .NET Framework (not .NET Core) that falls back to ANSI (which is definitely not appropriate - but I doubt they'll fix that).

* .NET Core now returns an encoding object with code page `0`  and web name `Codepage - 0`- I don't know if that's by design or a bug, but it certainly isn't helpful. 
  * By contrast, if you call `Encoding.GetEncoding(0)` explicitly, you get UTF-8, which is consistent with `Encoding.Default` in .NET Core - which makes sense, given that this call in isolation isn't console-related.
  * As @jborean93 has found out, with the Windows code-page encoding provider registered, which apparently happens automatically when a console _is_ attached (in a console application), it preempts the default meaning of `0` and returns the system's ANSI code page instead.

**Comment by mklement0** [comment]
@jborean93, the explanation for the discrepancy between `Invoke-Command` and `Enter-PSSession` is therefore as follows:

`Invoke-Command` doesn't call `[System.Text.Encoding]::RegisterProvider([System.Text.CodePagesEncodingProvider]::Instance)`, whereas `Enter-PSSession` does:

If register the code-page encoding provider explicitly, you get the same symptom as with `Enter-PSSession`:

```powershell
# !! Without the `[System.Text.Encoding]::RegisterProvider([System.Text.CodePagesEncodingProvider]::Instance)`
# !! call, the output is (0, 0)
PS>  icm -cn . -config PowerShell.7.2.0-preview.4 { [System.Text.Encoding]::RegisterProvider([System.Text.CodePagesEncodingProvider]::Instance); [console]::InputEncoding.CodePage; [console]::OutputEncoding.CodePage }
1252
1252 
```

Curiously, running _any_ console application in lieu of registering the code-page provider manages to make `Console.InputEncoding` - only - reflect the system's ANSI code page:

```powershell
PS>  icm -cn . -config PowerShell.7.2.0-preview.4 { 
  cmd /c exit
  [console]::InputEncoding.CodePage; [console]::OutputEncoding.CodePage
}
1252 # Console.InputEncoding
0    # Console.OutputEncoding
```

What it seems to come down is that calling the underlying WinAPI functions at least once  - even though they themselves keep returning `0`(!) - then makes `Console.In/OutputEncoding` return the system's active ANSI code page:

```powershell
PS>  icm -cn . -config PowerShell.7.2.0-preview.4 { 
  Add-Type -Namespace Util -Name WinApi -MemberDefinition '[DllImport("Kernel32.dll")] public static extern uint GetConsoleCP(); [DllImport("Kernel32.dll")] public static extern uint GetConsoleOutputCP();'; 
  [Util.WinAPI]::GetConsoleCP(); [Util.WinAPI]::GetConsoleOutputCP()
  [console]::InputEncoding.CodePage; [console]::OutputEncoding.CodePage
}
0    # GetConsoleCP()
0    # GetConsoleOutputCP()
1252 # Console.InputEncoding
1252 # Console.OutputEncoding
```

**Comment by mklement0** [comment]
As for what PowerShell should do:

It seems like the _caller's_ settings should determine the behavior:

* In terms of _culture_ (`[cultureinfo]::CurrentCulture` / `$PSCulture`), the _caller's_ settings already apply, irrespective of the target-machine settings.

* It therefore also makes sense to use the caller's _console encoding_.

However, that is problematic in _cross-platform_ remoting: 

* Windows -> Unix: The Windows code page may not even be available on the target machine.
* Unix -> Windows: This would (most likely) make the console use UTF-8

Generally, agreeing on UTF-8 as the "lingua franca" would make the most sense - certainly more than the current Windows behavior of using the _ANSI_ code page.

**Comment by iSazonov** [comment]
@dhowett Have you any thoughts?

**Comment by iSazonov** [comment]
> . I just don't have much of an idea of what the purpose of this particular code is for. It looks like some sort of shim to implement some .NET Framework behaviour that is missing in .NET Core. This could even be a red herring and not at all related to the problem but I thought it interesting enough to share.

Before we added the "Encoding.RegisterProvider(CodePagesEncodingProvider.Instance); " users could not use OEM and ANSII encoding names in cmdlets (Get-Content -Encoding windows-1251)

**Comment by iSazonov** [comment]
> `Invoke-Command` doesn't call `[System.Text.Encoding]::RegisterProvider([System.Text.CodePagesEncodingProvider]::Instance)`, whereas `Enter-PSSession` does:

If I understand right  `Invoke-Command` should follow `Enter-PSSession` behavior and both should have correct default that should be rather Utf8 then OEM in modern Windows 10 (I would not worry about old Windows 7). And we need a fix in Console class (no workaround could be in PowerShell itself?). Yes?

**Comment by mklement0** [comment]
>  users could not use OEM and ANSII encoding names in cmdlets (Get-Content -Encoding windows-1251)

Indeed; `[System.Text.Encoding]::RegisterProvider([System.Text.CodePagesEncodingProvider]::Instance)` is the prerequisite for access to all codings based on _Windows code pages_, notably including the active OEM and ANSI code pages; that is, without registering the provider, something like `Encoding.GetEncoding(437)` (the [OEM code page used in the `en-US` culture](https://en.wikipedia.org/wiki/Code_page_437), for instance) wouldn't work.

> both should have correct default that should be rather Utf8 then OEM in modern Windows 10 

Agreed, but to that end we should also commit to making  regular, local console windows default to UTF-8 - which is long overdue - as originally proposed in  #7233 and fleshed out in https://github.com/PowerShell/PowerShell/issues/14945#issuecomment-800503184

The only thing that could break is calls to very old CLIs that cannot handle code page `65001` (UTF-8) and its variable-length encoding aspect - but my guess is that such CLIs are pretty rare these days.

---

> And we need a fix in Console class (no workaround could be in PowerShell itself?). Yes?

No, if the intent is to do "our own thing" with unconditional use of UTF-8, I don't think we need a fix in `Console` (as debatable as code page `0` in the _absence_ of the Windows code-page encoding provider is, and as debatable as use of the ANSI code page is in its _presence_, but that's a separate issue).

Instead, we should do what @jborean93 has mentioned as an option [above](https://github.com/PowerShell/PowerShell/issues/15128#issuecomment-810932594): 

* Allocate a console on entering the remote session and set its encoding to UTF-8, and which is kept around for the lifetime of the session for all console applications (if any) to run in.
  * I think doing so on _entering_  rather than _on demand_ is preferable, so that someone executing `[Console]::OutputEncoding`, for instance,  _before_ calling a console application in the session gets the correct information.

**Comment by iSazonov** [comment]
> Agreed, but to that end we should also commit to making regular, local console windows default to UTF-8 - which is long overdue - as originally proposed in #7233 and fleshed out in [#14945 (comment)](https://github.com/PowerShell/PowerShell/issues/14945#issuecomment-800503184)
> 
> The only thing that could break is calls to very old CLIs that cannot handle code page `65001` (UTF-8) and its variable-length encoding aspect - but my guess is that such CLIs are pretty rare these days.

Yeah, hello Oracle! :-))) I guess there are a lot of old business applications using OEM encoding.
When I looked at this earlier I was thinking about the Native cmdlet concept to be able to annotate native applications accurately and personally.

**Comment by jborean93** [comment]
Even if there is an older legacy API the caller can still set `[Console]::OutputEncoding` to something else to change the code page.

**Comment by mklement0** [comment]
@iSazonov, note that use of OEM encoding _per se_ isn't the problem, as long as:

* programs do not use a _fixed_ OEM encoding and instead - as they should - respect the console's encoding (code page), which defaults to _system's_ OEM code page, as determined by the legacy system locale, aka "Language for non-Unicode programs), such as `437` on US systems.

* _and_ such programs know how to handle  the case when the console code page happens to be `65001`, i.e. UTF-8 (older programs may not be able to deal with the variable-length encoding aspect of UTF-8 (1-4 bytes per character)).

In short: If we switch PowerShell consoles from the system's OEM code page to `65001` == UTF-8, _only_ the following cases will amount to a breaking change:

* Legacy programs that use a _fixed_ OEM code page (other than `65001`).
* Legacy program that respect the console's code page but _cannot handle UTF-8_.

For "rogue" programs that do their own thing (e.g., Node.js, Python, `sfc.exe` - though note that Python can be configured to use UTF-8), the temporarily-setting-`Console]::OutputEncoding` workaround is _already_ necessary.

**Comment by TravisEz13** [comment]
Remoting working group:  The issues was filed with a description of a behavioral difference with no description of why the desired behavior is needed, what the scenario would be.  We understand that ideally this would be something else, but it is not clear from the issue what the behavior should be.

**Comment by jborean93** [comment]
It's simply someone opens up a remote PSSession and runs an external command that returns non-ASCII characters. Doing so will mean the string is incorrect when it is read by PowerShell because of the codepage difference. Not sure why you would close the issue as this is behaviour in PowerShell that should be corrected.

**Comment by jborean93** [comment]
![image](https://user-images.githubusercontent.com/8462645/113937701-e907e300-983c-11eb-9ebb-621d82a6e023.png)

Very simple reproducer that shows when a native application writes non-ASCII characters the values are screwed up, `é` becomes `‚` due to the encoding mismatch.

**Comment by jborean93** [comment]
If PowerShell was to run `AllocConsole` when it starts up when there is no console present we can see the correct behaviour returns back to us

![image](https://user-images.githubusercontent.com/8462645/113938225-a1ce2200-983d-11eb-896f-c59faf1a7221.png)

You don't even need to change the codepage to utf-8 by default, although I would recommend doing that as a follow up PR due to other reasons as stated in https://github.com/PowerShell/PowerShell/issues/14945.

**Comment by mklement0** [comment]
@TravisEz13, /cc @SteveL-MSFT:

> The issues was filed with a description of a behavioral difference with no description of why the desired behavior is needed

A reasonable expectation to have of a working group is a deep understanding of the issues involved, so that not every detail needs to be spelled out; thankfully, @jborean93 has since done so, leaving no doubt that things are _broken_.

>   it is not clear from the issue what the behavior should be.

See my previous point.

> seems like the only clear issue is on .NET

No, that issue is incidental, as [explained above](https://github.com/PowerShell/PowerShell/issues/15128#issuecomment-812681768).

> you can work around this by setting up the remote console how you want it to be. 

No, you can't, and it should have been obvious from the previous comments that you cannot (leaving impractical solutions involving ad-hoc compiled C# code that calls the Windows API to allocate a console aside).

In short: The working group's cursory and dismissive response is deeply concerning; it is something I first got a taste of in https://github.com/PowerShell/PowerShell/issues/11698, but Steve [assured me that the experience wasn't representative of what working groups would bring to the table](https://twitter.com/Steve_MSFT/status/1372657543747489792?s=20).
Certainly, this experience leaves me unconvinced.

**Comment by iSazonov** [comment]
> The trouble with the 2nd point and identifying the codepage that the subprocess will use is a reason why I think PowerShell itself should just spawn it's own console when it starts (or when the first process is spawned). The first option also has the added benefit of allowing PowerShell to set the codepage to UTF-8 in the future in case that's something to be desired.

Is it always permissible to create a console? Are there any unwanted side effects?

**Comment by TravisEz13** [comment]
> A reasonable expectation to have of a working group is a deep understanding of the issues involved, so that not every detail needs to be spelled out; thankfully, @jborean93 has since done so, leaving no doubt that things are broken.

It's reasonable to have issue filed with the required information which include what you are trying to accomplish.  Issues without that information, are **NOT** issues at all but merely discussions and not actionable by the working group.  The primary action of the working group is to prioritize an issue.  An issue without a scenario has no priority and in an area as complex as remoting, I don't believe we (Including the community) should fix any issue like that.

**Comment by TravisEz13** [comment]
In reply to 
> The working group's cursory and dismissive response is deeply concerning

and

> It's simply someone opens up a remote PSSession and runs an external command that returns non-ASCII characters. Doing so will mean the string is incorrect when it is read by PowerShell because of the codepage difference. Not sure why you would close the issue as this is behaviour in PowerShell that should be corrected.

This was not put in the issue description.  This type of information should be included in the issue description to help the working group understand the impact of the issue.  If it is discovered later, please update the initial issue.

**Comment by mklement0** [comment]
@TravisEz13 

* While the _full_ picture only emerged in the course of the discussion, _it should be obvious that the behavior is broken_, so the answer to "what are you trying to accomplish" is: _fix broken behavior_.

* Perhaps we need clarity on the duties of working groups; here's what I think they should _not_ do: Close issues:
    * without having taken the time to understand the issue at hand. 
    * based on arguments that do not apply ("you can work around this by setting up the remote console how you want it to be")

* If the problem isn't fully understood yet and / or you don't want to spend the time gleaning it from the discussion, _ask for clarification or an update_ (as you now have), or even _ask for a new, focused proposal / bug report to be created_.

  * I can appreciate the problem of not wanting to wade through a lengthy discussion, but to dismiss it summarily without even having understood the problem at hand is the disconcerting part.

  * If working groups routinely behave the way they did in this case, we've made no progress: it is reminiscent of similar committee decisions, where - understandable - a stretched-way-too-thin, small group of individuals had to make _all_ the decisions. My hope was that working groups would address that very problem: subject-matter experts that have the time to engage in relevant discussions (which, again,  doesn't preclude asking for a tl;dr first).

In other words: 

* Yes, this is an _ongoing_ discussion. 
* It even led to a [concrete, focused proposal for how to fix the issue](https://github.com/PowerShell/PowerShell/issues/15128#issuecomment-812681768).
* No one asked for the discussion _to end yet_, if there's no consensus yet or if the issue must be further clarified.

> This was not put in the issue description.

Is it really official policy that the only thing working groups are required to read is the _issue description_?
I generally try to keep the initial post updated, but in an ongoing discussion that's not always practical.

My understanding - and hope - is that working groups are meant to _work with the community_ rather than handing down 2-minute-deliberation fiat decisions.

By contrast, if the intent is indeed to have working groups only come in _once a discussion has settled down and a focused proposal has been made_, we need a different process: currently, issue are tagged for the relevant workgroup _shortly after creation_, so there's no control over the stage of the discussion at which the working group comes in.

**Comment by mklement0** [comment]
@iSazonov 

> Is it always permissible to create a console? Are there any unwanted side effects?

As far as I know, always creating a console should be fine, and I'm not aware of any side effects, but perhaps @jborean93 has further insights.

> Will this be enough to solve this problem, or is something else absolutely necessary?

Yes, allocating a console at the start of the session should be enough to solve the problem:

* Even without switching to UTF-8, it restores expected console behavior - albeit based on the _target_ machine's active OEM code page when targeting a Windows machine.

  * That said, I fully agree with @jborean93 that defaulting PowerShell consoles to UTF-8 is the right thing to do.

* Once a console is available, `[Console]::OutputEncoding` can be set (temporarily) as usual in order to handle nonstandard encoding behavior, such as the always-UTF-8 output from Node.js applications.

Here's an example that emulates the proposed fix, similar to @jborean93's example:

```powershell
# Target a Windows machine.
# E.g., from that machine itself, from an ELEVATED session:
PSELEV> Invoke-Command -cn . {
  # Allocate a console for the lifetime of this session. THIS IS WHAT POWERSHELL SHOULD DO AUTOMATICALLY.
  $null = (Add-Type -Name Aux -Namespace WinApi -Passthru -MemberDefinition '[DllImport("kernel32.dll")] public static extern bool AllocConsole();')::AllocConsole()
  # Now you can set [Console]::OutputEncoding to control how external-program output is decoded:
  # E.g., Node.js always output UTF-8
  # !! Without a console, this assignment would FAIL.
  [Console]::OutputEncoding = [System.Text.UTF8Encoding]::new()
  # Now Node.js output is captured correctly.
  ($output = node -pe "'eé'")
}

eé  # OK
```

Note that in the "loopback" remoting scenario you actually do see the allocated console actually becomes visible, in mangled form - I haven't looked into how to prevent that.

**Comment by iSazonov** [comment]
> > Is it always permissible to create a console? Are there any unwanted side effects?
> 
> As far as I know, always creating a console should be fine, and I'm not aware of any side effects, but perhaps @jborean93 has further insights.

I'd want to get confirmation from @TravisEz13 that we do not fall in a security issue here.  (I vaguely remember an old story with shadow window for system account  on Windows)

**Comment by mklement0** [comment]
> Do we want to fix both scenario - Enter-PSSession and Invoke-Command?

Absolutely - calling console applications should be supported in both scenarios.

**Comment by iSazonov** [comment]
@mklement0 We have an open PR in RFC repo for WG - please move your feedback to the PR.

**Comment by iSazonov** [comment]
@mklement0 After we get a confirmation for my question https://github.com/PowerShell/PowerShell/issues/15128#issuecomment-816373098 I suggest you to open new issue with clean and short description (with references to posts in that issue) and proposed solution. Thanks you and @jborean93 for great discussion!

**Comment by mklement0** [comment]
Thanks, @iSazonov.

As for the security question: note that consoles already are being created if you call console applications, namely _transient_ ones, on a _per-call_ basis - but with misinterpreted-by-PowerShell output, and with no way to fix that (except via the WinAPi call shown above).

**Comment by iSazonov** [comment]
> note that consoles already are being created if you call console applications

The consoles are created by OS for subprocesses. My question is about root WinRM process.

**Comment by iSazonov** [comment]
This can have a side effect #15229

**Comment by jborean93** [comment]
> This can have a side effect #15229

In this cases these 2 issues are unrelated and is no indication that allocating a console for WSMan is going to be a problem. For WSMan the PSRP exchanges occur on the WSMan transport (see SOAP/HTTP exchange) which is totally unrelated to the console stdio pipes. In the case of SSH there is already a console allocated which is fine but the problem is that the existing stdio referenced in PowerShell is now the sharing the same pipe as what PowerShell is using to exchange the PSRP messages. So by writing to `stdout` you are sending a message back to the client which isn't a correct PSRP packet causing a failure. For WSMan by writing to the allocated console nothing happens, it's just written to the conhost buffer that was allocated and not seen anywhere which is exactly what happens today with no console being there.

**Comment by PaulHigin** [comment]
wg-remoting

@mklement0  It looks like this has evolved over time through this discussion, and I am confused as to the final problem.  It is not clear to me if there is an issue in PowerShell Core or PowerShell Windows or both.  Can you please update this issue (at the beginning and including issue title) for clarity and provide a summary of the problem including scenario based repro?

To clarify, I realize that remoting session does not set the console encoding, but I am unclear of a scenario where this causes a problem.

**Comment by jborean93** [comment]
@PaulHigin I can't update the title but essentially it's a problem in both WinPS and PSCore, the end result just might be different even if they are still both wrong.

Essentially for WSMan based remoting sessions there is no `conhost.exe` assigned to the process so when PowerShell goes to determine the encoding that is used to encode the bytes received from a native app invocation it will most likely use the wrong encoding. The reason why it will most likely be the wrong encoding is because:

* The native application is spawned with a new console (there's no console on the ps process to inherit from) and the default output encoding is most likely to be the console codepage
  * The only exceptions to this rule are things like `python`, `node` which don't respect the console codepage settings and use their own logic
  * `python`, `node`, and others are still a problem but unrelated to this particular issue
* When PowerShell reads the raw bytes on that native apps `stdout` pipe it essentially uses `[Console]::OutputEncoding`
  * This may in fact be logic in .NET but the logic is still the same
* `[Console]::OutputEncoding` somewhat "fails" because the remote PowerShell session does not have an actual console associated with it
  * It doesn't actually throw an exception but fails because there is no console in the ps session to get the output encoding values for
  * The underlying Win32 API will return a codepage of `0` in this case
  * This typically corresponds to the "ANSI" codepage which is **not** the same as the console codepage
  * This behaviour might be an issue but it would be up to dotnet to decide if they want to change that behaviour
  * Even if they do I don't see a nice way to solve it as there is some complex logic that decides what the console codepage will be
* Because the output is decoded using the "ANSI" codepage but the native app encoded it using the console codepage we get the wrong string value in PowerShell

To test this out you can run the following on either WinPS or PSCore.

```powershell
$cmd = {
    [Console]::Out.WriteLine('café')
}
$encCommand = [Convert]::ToBase64String([Text.Encoding]::Unicode.GetBytes($cmd.ToString()))

# This works because [Console]::OutputEncoding is set for a normal process
powershell.exe -EncodedCommand $encCommand

# This doesn't work because the remote process started by the WSMan plugin doesn't have a console process
'Microsoft.PowerShell', 'PowerShell.7' | ForEach-Object -Process {
    $session = New-PSSession -ComputerName localhost -ConfigurationName $_
    try {
        Write-Host "The output with Invoke-Command PSSession $_"
        Invoke-Command $session -ScriptBlock {
            powershell.exe -EncodedCommand $using:encCommand
        }

    }
    finally {
        $session | Remove-PSSession
    }
}
```

For me this outputs `caf‚` because PowerShell will output `café` using the 437 codepage (`0x63 0x61 0x66 0x82`) but PowerShell is reading it using my "ANSI" codepage (windows-1252); `0x82` in [windows-1252](https://en.wikipedia.org/wiki/Windows-1252) is `‚`. Note the string iterpretation happens on the remote side so this isn't a CLIXML serialization over PSRemoting but how PowerShell itself decodes bytes from a native app if there is no console associated with the PowerShell process.

My proposed fix for this in PowerShell is to always ensure that a console is associated with the process so that `[Console]::OutputEncoding` actually has a correct value just like it does for a native app. I'm sure there are some edge cases that need to be tested but as per https://github.com/PowerShell/PowerShell/issues/15128#issuecomment-815282724 it does fix the issue.

**Comment by mklement0** [comment]
Thank you, @jborean93, for succinctly summarizing the issue.

Re:

> When PowerShell reads the raw bytes on that native apps stdout pipe it essentially uses `[Console]::OutputEncoding`
> * This may in fact be logic in .NET but the logic is still the same

It is actually _PowerShell's_ decision (see #10824), albeit a sensible one.

Two asides: 

* The behavior deserves documenting, as suggested in the proposed `about_Native_Calls` conceptual help topic - see https://github.com/MicrosoftDocs/PowerShell-Docs/issues/5152

* Defaulting _all_ console contexts to UTF-8 is called for: see https://github.com/PowerShell/PowerShell/issues/14945#issuecomment-800503184

---

@PaulHigin, if you'll excuse the mixed metaphor, I'm [personally tired of shouting at windmills](https://github.com/PowerShell/PowerShell/pull/10238#issuecomment-830406280), so I'm unsubscribing from this thread.

I'll leave any follow-up in the capable hands of @jborean93, who - I hope - has the stamina to keep going.

**Comment by PaulHigin** [comment]
@jborean93  Thanks for the clarification.
I feel the title for this issue should be:
**WinRM remote sessions do not set Console code page to that of client, affecting remote native command execution**
@TravisEz13 I don't have permissions to update the Issue title, can you do it?

I feel a better repro scenario would be something like:
```powershell
# Create WinRM session from client with non-default codepage
# Run script executing native command in remote session

Expected Results:
"text"

Actual Results:
"garbled text"
```

@jborean93, @mklement0 
The first thing the working group does is triage the issue, and this means understanding the issue and trying to determine user impact for prioritization.  So it is critical to have this information at the very beginning and be as clear and concise as possible.  Discussion, and proposed solutions are very welcome but only after the issue is clearly defined.   In this case the issue was more like a discussion but without any clear conclusion summary.  This is why guidelines require a specific format ... to help working groups quickly understand impact and triage issues.  In hindsight, I feel this should have started as a discussion, eventually culminating into an Issue once the problem was fully understood.

**Comment by jborean93** [comment]
> I feel the title for this issue should be: WinRM remote sessions do not set Console code page to that of client, affecting remote native command execution

@PaulHigin that is not the problem here, strings returned from a remote session are serialized in CLIXML using utf-8 and aren't a problem here. The remote session also does not need to match with the console code page of the client, in fact you don't want that as it will exacerbate the problem in case they do mismatch.

The issue is the WinRM remote session doesn't have a console associated with it so the `[Console]::OutputEncoding` value isn't correct. This causes the problem where any spawned console processes will use the actual console codepage but PowerShell is using a different encoding when getting the output causing the mismatch values for non-ASCII characters.

> I feel a better repro scenario would be something like:

If you really think so, personally I don't as you can't see what script can actually replicate this problem, at least with mine it showed you the problem exists on both WinPS and PSCore and how to replicate this exact problem.

**Comment by PaulHigin** [comment]
@jborean93  Thanks again for the clarification.  What I am looking for is real world impact of a lack of correct console code page in a remote session.  Why would anyone use console write APIs in a remote script?  If you can provide a real world example that will help us assess impact.  But at this point it seems to be a low impact issue.  Sorry if I am missing something obvious.

**Comment by jborean93** [comment]
> What I am looking for is real world impact of a lack of correct console code page in a remote session

The impact is whenever a native application outputs a non-ASCII characters (0-127), there's a very high chance that because of the encoding mismatch the string stored in PowerShell is not the same as the one the native application outputted.

> Why would anyone use console write APIs in a remote script?

They wouldn't, or if they do PowerShell won't capture it in any scenario (even locally) because it bypasses the stdout pipe and writes directly to the screen buffer. On the remote side this screen buffer is not the local one so it is lost to the world, locally this bypasses the PowerShell stdout capturing and the console codepage doesn't apply. This is unrelated to the problem and not something that PowerShell can really fix unless it starts delving into the ConPTY stuff that WT uses.

> If you can provide a real world example, that will help us assess impact.

It's hard to as for myself personally I rarely deal with non-ASCII characters, most of my tools is based on en-US and these diacritics isn't very common to encounter in native tools. Essentially any native app that outputs a non-ASCII character will most likely be affected.

One other example where this is probably even more problematic is the `tree` command someone might be doing to get a snapshot of what a directory looks like:

```
# Run on a WSMan remote PSSession
> tree 'C:\Program Files\WindowsPowerShell\Modules'
```

```
Folder PATH listing for volume Windows 2019
Volume serial number is 08B0-E145
C:\PROGRAM FILES\WINDOWSPOWERSHELL\MODULES
ÃÄÄÄImpliedReflection
³   ÀÄÄÄ0.2.1
³       ÀÄÄÄen-US
ÃÄÄÄMicrosoft.PowerShell.Operation.Validation
³   ÀÄÄÄ1.0.1
³       ÃÄÄÄDiagnostics
³       ³   ÃÄÄÄComprehensive
³       ³   ÀÄÄÄSimple
³       ÀÄÄÄTest
³           ÀÄÄÄModules
³               ÃÄÄÄExample1.Diagnostics
³               ³   ÀÄÄÄDiagnostics
³               ³       ÀÄÄÄSimple
³               ÃÄÄÄExample2.Diagnostics
³               ³   ÀÄÄÄ1.0.1
³               ³       ÀÄÄÄDiagnostics
³               ³           ÀÄÄÄSimple
³               ÀÄÄÄExample3.Diagnostics
³                   ÃÄÄÄ1.1.1
³                   ³   ÀÄÄÄDiagnostics
³                   ³       ÀÄÄÄSimple
³                   ÀÄÄÄ2.0.1
³                       ÀÄÄÄDiagnostics
³                           ÀÄÄÄSimple
ÃÄÄÄPackageManagement
³   ÀÄÄÄ1.0.0.1
³       ÃÄÄÄDSCResources
³       ³   ÃÄÄÄen-US
³       ³   ÃÄÄÄMSFT_PackageManagement
³       ³   ³   ÀÄÄÄen-US
³       ³   ÀÄÄÄMSFT_PackageManagementSource
³       ³       ÀÄÄÄen-US
³       ÀÄÄÄen
ÃÄÄÄPackerWindoze
ÃÄÄÄPester
³   ÀÄÄÄ5.1.1
³       ÃÄÄÄbin
³       ³   ÃÄÄÄnet452
³       ³   ÀÄÄÄnetstandard2.0
³       ÀÄÄÄen-US
ÃÄÄÄPInvokeHelper
³   ÀÄÄÄ0.1.0
ÃÄÄÄplatyPS
³   ÀÄÄÄ0.14.1
³       ÃÄÄÄdocs
³       ÃÄÄÄen-US
³       ÀÄÄÄtemplates
ÃÄÄÄPowerShellGet
³   ÀÄÄÄ1.0.0.1
³       ÀÄÄÄen-US
ÃÄÄÄpsprivilege
³   ÀÄÄÄ0.1.0
³       ÃÄÄÄPrivate
³       ÀÄÄÄPublic
ÀÄÄÄPSReadline
    ÀÄÄÄ2.0.0
        ÀÄÄÄen
```

As you can see the characters that `tree` uses to denote the structure are all non-ASCII chars and are affected by the wrong encoding.

Here is some details on those non-ASCII chars used in the output

|437 Char|437 Decimal|Windows-1252 Char|
|-|-|-|
|`│`|179|`³`|
|`└`|192|`À`|
|`─`|196|`Ä`|

> But at this point it seems to be a low impact issue

That's understandable, my intention is to just really document the problem so that either;

1. People can understand what is happened - obviously I've failed at that with the numerous posts
2. A workaround if no fix has been merged into PowerShell - Call `AllocConsole` using PInvoke

**Comment by PaulHigin** [comment]
This is exactly what I said before is it not?  A native console command that returns text and runs in a remote session may return garbled text if the console code page is incorrect.  Right?  If so, we can update the Issue title and proceed with triage.

**Comment by jborean93** [comment]
My apologies I misinterpreted it but after rereading it does sound like we are on the same page. I can open another issue under my name if we want to start without all this history here with a different title if you wish.

**Comment by PaulHigin** [comment]
Thanks @jborean93 , I think the history is valuable so let's keep it.  I just wanted a clearer understanding of the problem.  We'll update this issue at the next working group meeting.

**Comment by SteveL-MSFT** [comment]
First, thanks @mklement0 and @jborean93 for the detailed investigation.  A couple of questions:

- Is this only a problem with WinRM based remoting or is SSH also equally affected?
- Is the ask to have pwsh.exe call AllocConsole() on startup?  Note that we are not making changes to powershell.exe that aren't security or other compliance issues

**Comment by jborean93** [comment]
> Is this only a problem with WinRM based remoting or is SSH also equally affected?

Just WinRM, the process that is spawned by the WinRM service seems to be done in a way that there's no console associated with it (`wsmprovhost.exe`). SSH calls `pwsh.exe` like a normal process and has a `conhost` associated with it so isn't affected.

> Is the ask to have pwsh.exe call AllocConsole() on startup?

Or at least allocate it if there is no console that exists already. This at least solves the issue in testing but I'm unsure if it will start to affect things like the `$host.UI.WriteLine` behaviour or not. I assume it wouldn't but it's hard to tell. I also don't know if this should be done in `pwsh.exe` as a WinRM based remote session doesn't seem to call that executable. It uses `wsmprovhost.exe` and `pwrshplugin.dll` so maybe that's what needs to be touched.

> Note that we are not making changes to powershell.exe that aren't security or other compliance issues

That's understandable, the comparisons with WinPS and PSCore are mostly to show it's an existing problem and not something new added since WinPS.

**Comment by SteveL-MSFT** [comment]
Maybe changing `pwsrhplugin.dll` is the way to go.  That code is [here](https://github.com/PowerShell/PowerShell-Native/tree/master/src/powershell-native/nativemsh/pwrshplugin).

@jborean93 would you be willing to build a private validating that `AllocConsole()` solves the problem?

If we agree the change is in that dll, I'll move this issue over to that repo.

**Comment by jborean93** [comment]
I can try but I can't guarantee I'll be able to look at where it needs to go and build it anytime soon.

**Comment by SteveL-MSFT** [comment]
@jborean93 take a look at the referenced PR.  Did some manual validation with `tree` and your encoded command example and works correctly.

**Comment by jborean93** [comment]
Thanks for the PR, I've been able to test it and can also confirm it fixes this particular problem. I also couldn't find any side affects and things acted like they did before.

**Comment by jborean93** [comment]
Thanks all for sticking to this and working on a solution.

**Comment by SteveL-MSFT** [comment]
@jborean93 your detailed investigation made this easy :)

**Comment by TravisEz13** [comment]
@jborean93 Thanks for working with @PaulHigin to help the Working group and PowerShell understand the issue.  This is vital part of the process.
